src/Backups/BackupCoordinationLocal.h

```diff
@@ -48,17 +48,18 @@ class BackupCoordinationLocal : public IBackupCoordination
 
     std::optional<FileInfo> getFileInfo(const String & file_name) const override;
     std::optional<FileInfo> getFileInfo(const SizeAndChecksum & size_and_checksum) const override;
-    std::optional<SizeAndChecksum> getFileSizeAndChecksum(const String & file_name) const override;
 
     String getNextArchiveSuffix() override;
     Strings getAllArchiveSuffixes() const override;
 
 private:
+    bool deduplicate;
     mutable std::mutex mutex;
     BackupCoordinationReplicatedTables replicated_tables TSA_GUARDED_BY(mutex);
     BackupCoordinationReplicatedAccess replicated_access TSA_GUARDED_BY(mutex);
     std::map<String /* file_name */, SizeAndChecksum> file_names TSA_GUARDED_BY(mutex); /// Should be ordered alphabetically, see listFiles(). For empty files we assume checksum = 0.
     std::map<SizeAndChecksum, FileInfo> file_infos TSA_GUARDED_BY(mutex); /// Information about files. Without empty files.
+    std::map<String /* file_name */, FileInfo> file_names_infos TSA_GUARDED_BY(mutex); /// Should be ordered alphabetically, see listFiles(). For empty files we assume checksum = 0.
```

Comment:
Now we're going to have one more map in addition to those two maps we already have. It doesn't seem all those changes in `BackupCoordination*` files are really necessary. It looks like all you had to do is to ignore `is_data_file_required` in `BackupImpl` after

```
coordination->addFileInfo(info, is_data_file_required);
```

and recalculate it after calling `addFileInfo()` like this:

```
if (!deduplicate && info.size())
  is_data_file_required = true
```
and that's all!

Reply:
>It looks like all you had to do is to ignore is_data_file_required in BackupImpl.cpp

Yes, it should work, however I though that you will like the patch with separate map more, since it does not play games with flags.

Actually looking into this a little bit more and it seems that this cannot be done, since in this case meta information will be wrong, at least `pos_in_archive`/`archive_suffix` since it stored only there and can differs for duplicated files.

Reply:
Though looks like `pos_in_archive` is never used/set..., but anyway there is still `archive_suffix`.

Reply:
`archive_suffix` is also never used. It's a part of a planned solution which would allow to write a backup a sequence of *.zip archives, but it has not finished and right now it's just disabled.

Anyway I'd like to avoid injecting this deduplication feature deeply inside BackupCoordination classes.

In fact this deduplication feature can be completely implemented by changing only the function `BackupImpl::finalizeWriting()` - you can just write backup as usual, and then before writing its metadata you can check which files were deduplicated and write them too. This approach can be even better - because you don't really need to upload files which are already on S3 - you can copy them inside S3 instead.

Reply:
That can be made as following:
1) `IBackupWriter` has the function `copyFileNative()` - it was designed to copy from DiskS3, but it seems it can be modified to allow copying a random object from S3 to S3.
2) Before calling `backup->finalizeWriting()` the class `BackupWorker` can get `FileInfos` from `BackupImpl` and copy duplicated files using `IBackupWriter` and the thread pool `backups_thread_pool`, then pass some flag to `backup->finalizeWriting()` so `BackupImpl::writeBackupMetadata()` would write FileInfos in deduplicated manner (i.e. without "data_file" keys)

Reply:
The idea of using `CopyObjectRequest` looks promising to avoid writing to S3 those files again. Thought likely all deduplicated files will be small (so as their total size).

>IBackupWriter has the function copyFileNative() - it was designed to copy from DiskS3, but it seems it can be modified to allow copying a random object from S3 to S3.

Do you suggest to store everything that is required for native copy (all objects path and credentials)?
Actually if each shard will do it by itself the disk can be stored and no need to change interface.

>Before calling backup->finalizeWriting() the class BackupWorker can get FileInfos from BackupImpl and copy duplicated files using IBackupWriter and the thread pool backups_thread_pool.

But you also need to update the information in coordinator anyway, because otherwise metadata (`.backup`) will be wrong.

Reply:
> But you also need to update the information in coordinator anyway, because otherwise metadata (.backup) will be wrong.

The backup will be still valid and can be restored in its usual way, so why is that wrong?